src/libsyntax/feature_gate.rs

```diff
@@ -718,6 +719,41 @@ pub enum AttributeGate {
     Ungated,
 }
 
+#[derive(Clone, Copy)]
+pub struct AttributeTemplate {
+    word: bool,
+    list: Option<&'static str>,
+    name_value_str: Option<&'static str>,
+}
+
+impl AttributeTemplate {
+    fn compatible(&self, meta_item_kind: &ast::MetaItemKind) -> bool {
```

Comment:
I'd like a simple doc-comment, maybe:

"Check that the given attribute is compatible with this template"